Serialize NSAttributedString font attributes as WebCore::Font

achristensen07 opened the pull request:
#### 493b7f1a6f9d8f78fbd136ce221eb7c14b21d4d0
<pre>
Serialize NSAttributedString font attributes as WebCore::Font
<a href="https://bugs.webkit.org/show_bug.cgi?id=254860">https://bugs.webkit.org/show_bug.cgi?id=254860</a>
rdar://107503178

Reviewed by Myles Maxfield.

This also fixes a regression from 262192@main where if you make an NSAttributedString
from an HTML string that contains a web font as a data URL, we now get an attribute
with the key of NSFontAttributeName again.

* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:

Canonical link: <a href="https://commits.webkit.org/262556@main">https://commits.webkit.org/262556@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5226dbb0a67b490faf11b5cb3a9800ddd82347c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1936 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1916 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1704 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1842 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1650 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1633 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2594 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1617 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1543 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1680 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1660 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2754 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1701 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1514 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1626 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1624 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1775 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/197 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -76,13 +79,13 @@
-    }, [] (const RetainPtr<UIFont>& value) -> RetainPtr<NSObject> {
-        return value;
+    }, [] (const Ref<Font>& font) -> RetainPtr<NSObject> {
+        return (UIFont *)font->getCTFont();
```
We should use `bridge_cast` here (and if we can't use `bridge_cast` right now, we should fix it so that we can use it here).

litherum commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -172,7 +175,7 @@
         return { { { RetainPtr { textAttachment } } } };
     if (auto* font = dynamic_objc_cast<NSFont>(value))
-        return { { { RetainPtr { font } } } };
+        return { { { Font::create(FontPlatformData(toCTFont(font), font.pointSize)) } } };
```
Ditto

achristensen07 replied:
We would have to first cast from id to a UIFont, then call bridge_cast, which would require a bunch of platform-specific stuff to be added.  I'm going to use __bridge instead for this PR and if further cleanup is desired it can be done separately.

litherum commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -183,7 +186,7 @@
         return { { { RetainPtr { (NSTextAttachment *)value } } } };
     if ([value isKindOfClass:PAL::getUIFontClass()])
-        return { { { RetainPtr { (UIFont *)value } } } };
+        return { { { Font::create(FontPlatformData((CTFontRef)value, [(UIFont *)value pointSize])) } } };
```
The whole point of `PAL::getUIFontClass()` is so you don't cast to `UIFont` directly.

litherum commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
```diff
@@ -497,6 +497,7 @@
     return nullptr;
 }
 
+// FIXME: This is duplicate code.
```
@mattwoodrow just duplicated these like last week. Can we make sure he un-duplicates them soon?

litherum commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
```diff
@@ -510,10 +511,8 @@
     // We can fall back to doing our best to find it ourself.
     fontDescriptor = findFontDescriptor(referenceURL, desiredPostScriptName);
-    if (!fontDescriptor) {
-        ASSERT_NOT_REACHED();
```
Shouldn't we keep this?

achristensen07 replied:
No, this assertion is hit in my test.

achristensen07 committed:
Serialize NSAttributedString font attributes as WebCore::Font
https://bugs.webkit.org/show_bug.cgi?id=254860
rdar://107503178

Reviewed by Myles Maxfield.

This also fixes a regression from 262192@main where if you make an NSAttributedString
from an HTML string that contains a web font as a data URL, we now get an attribute
with the key of NSFontAttributeName again.

* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:

Canonical link: https://commits.webkit.org/262556@main
